### PR TITLE
Issue1215 Adjust  is_public calculation for mediafiles, calculate always

### PR DESCRIPTION
**Pull request by r-peschke**
Can be reviewed, migrations are ready. Still missing recreation with calculated field handlers

### CODE REVIEW

File: tests/system/action/mediafile/test_upload.py [hunk 1]
```diff
@@ -199,3 +166,157 @@ def test_create_added_mimetype_ttf(self) -> None:
...
+        self.media.upload_mediafile.assert_called_with(file_content, 11, "text/plain")
+        self.assert_model_exists(
+            "group/1",
+            {
+                "mediafile_access_group_ids": [11],
+                "mediafile_inherited_access_group_ids": [11],
+            },
+        )
+
+    def test_createX_media_directory_group(self) -> None:
```
Comment (jsangmeister) [inline tests/system/action/mediafile/test_upload.py:222]:
```suggestion
    def test_create_media_directory_group(self) -> None:
```

**Review by jsangmeister — APPROVED** [review]
looks good besides missing migrations. I'm not sure how a migration for this would be implemented the best - theoretically, you would need to recalculate the field (call the `calculated_fields_helper`) for each media file, top-down in the meeting's tree structure. This is hard to do in a migration. Maybe we rather provide a small script which does exactly this and can be called manually for instances where this problem occurs? It would then look like this:
- fetch all mediafiles from this instance with their `parent_id`
- rebuild the tree structure for each meeting
- go through each tree bottom-down, call `mediafile.update` for each mediafile with only the id as payload

This is basically also what would need to be done in a migration, but seems way harder to implement.

**Comment by jsangmeister** [comment]
> looks good besides missing migrations. I'm not sure how a migration for this would be implemented the best - theoretically, you would need to recalculate the field (call the `calculated_fields_helper`) for each media file, top-down in the meeting's tree structure. This is hard to do in a migration. Maybe we rather provide a small script which does exactly this and can be called manually for instances where this problem occurs? It would then look like this:
> 
> * fetch all mediafiles from this instance with their `parent_id`
> * rebuild the tree structure for each meeting
> * go through each tree bottom-down, call `mediafile.update` for each mediafile with only the id as payload
> 
> This is basically also what would need to be done in a migration, but seems way harder to implement.

Talked with @r-peschke and came to the conclusion that a separate script is not the correct solution, but a "one-shot" update is  better and less work then a "real" migration where all events are updated. Will look into the datastore code and check how easily this is possible.

**Comment by jsangmeister** [comment]
As discussed: A "real" migration is the way to go, "one-shot" migrations are not easily implementable

File: migrations/migrations/0019_calc_mediafile_is_public.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                self.mediafiles[id_] = self.new_accessor.get_model_ignore_deleted(
+                    event.fqid
+                )[0]
+            if remove_dict := event.remove:
+                for key, value in remove_dict.items():
+                    if not value:
+                        continue
+                    assert isinstance(
+                        old_value := self.mediafiles[id_].get(key), list
+                    ), f"'{event.fqid}' should have values for '{key}', because there is a ListUpdate.remove!"
```
Comment (jsangmeister) [inline migrations/migrations/0019_calc_mediafile_is_public.py:68]: This is wrong, see https://github.com/OpenSlides/OpenSlides/wiki/Datastore-Service#interface: If a `remove` happens on a non-existent field, it is simply ignored.

File: migrations/migrations/0019_calc_mediafile_is_public.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+
+def calculate_inherited_groups_helper(
+    access_group_ids: Optional[List[int]],
+    parent_is_public: Optional[bool],
+    parent_inherited_access_group_ids: Optional[List[int]],
+) -> Tuple[bool, List[int]]:
+    """
+    TODO: If migration and backend are combined one day,
+    import this function from openslides_backend/models/helper.py
```
Comment (jsangmeister) [inline migrations/migrations/0019_calc_mediafile_is_public.py:187]: Thought again about this and I think it's actually good to keep the migration separated from the actual code. So I would just remove this TODO.

File: openslides_backend/action/actions/group/delete.py [hunk 4]
```diff
@@ -14,3 +22,159 @@ class GroupDeleteAction(DeleteAction):
...
     permission = Permissions.User.CAN_MANAGE
+
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        instance = super().update_instance(instance)
+        group = self.datastore.get(
+            FullQualifiedId(Collection("group"), instance["id"]), []
+        )
+        self.mediafile_ids: List[int] = list(
+            (set(group.get("mediafile_access_group_ids", set())) or set())
+            | (set(group.get("mediafile_inherited_access_group_ids", set())) or set())
```
Comment (jsangmeister) [inline openslides_backend/action/actions/group/delete.py:33]: If I counted the brackets correctly, you write here `set(...) or set()`, so you should be able to remove the second `set()`

**Review by jsangmeister — CHANGES_REQUESTED** [review]
Did not check every test, but the gist of it looks good. Please go ahead and test this on actual meeting data:
1) Create a meeting and upload a mediafile (or multiple, as you wish) with the current backend main
2) Verify that cloning this meeting fails, since `is_public` is set wrong
3) Shutdown the backend and restart it with this PR
4) Migrate by executing `python migrations/migrate.py finalize`
5) Assert that you can clone the meeting now

Process described from the top of my head, may contain errors.

File: migrations/migrations/0019_calc_mediafile_is_public.py (hunk 3 above)
Comment (r-peschke) [reply migrations/migrations/0019_calc_mediafile_is_public.py:187]: done

File: openslides_backend/action/actions/group/delete.py (hunk 4 above)
Comment (r-peschke) [reply openslides_backend/action/actions/group/delete.py:33]: done

**Review by peb-adr — APPROVED** [review]
Ok, I tested this successfully now.
Tried provocating other errors by e.g. cloning the clone, removing and adding files in the clones and the original meeting with different users.
All seems stable to me now.

**Pull request merged**